File: Sort algorithms/Shell Sort.cpp
//
//  shell.cpp
//
//
//  Created by 诸葛俊伟 on 1/30/16.
//  Copyright © 2016 诸葛俊伟. All rights reserved.
//  Lastest modified on 2/1/2016.
//

#include <iostream>
#include <vector>
using namespace std;
//typedef vector<int> number;

void swap(int* x, int* y)
{
    int temp = *y;
    *x = *y;
    *y = temp;
}

void shell(int num[], int size)
{
    int gap, j, i;
    for (gap = size/2; gap > 0; gap /= 2)
        for (i = gap; i < size; ++i)
            for (j = i - gap; j >= 0 && num[j] > num[j+gap]; j -= gap)
            {
                swap(num[j], num[j+gap]);
            }
}

int main()
{
    int num[] = {3,2,1,5,4,6,1,10};
    int size = 8;
    cout<<"original array:"<<endl;
    for (int i = 0; i < size; ++i)
        cout<<num[i]<<" ";
    cout<<endl;
    shell(num, size);
    cout<<"shelled array:"<<endl;
    for (int i = 0; i < size; ++i)
        cout<<num[i]<<" ";
    cout<<endl;
    return 0;
}

File: Sort algorithms/Shell Sort.swift
import Darwin

// shell sort

func shell(inout arr: [Int], _ n: Int)
{
    var gap = n/2
    while gap > 0 {
        for i in gap..<n {
            var j = i - gap
            while j >= 0 && arr[j] > arr[j+gap] {
                swap(&arr[j], &arr[j+gap])
                j = j - gap
            }
        }
        gap /= 2
    }
}

var a = [1,2,1,4,7,6,0]
let n = a.count
shell(&a, n)